icepyx/core/hdf5_arrow_table.py

```diff
@@ -0,0 +1,181 @@
+        self.credentials = credentials
+        self.mask_using = mask_using
+
+        self.h5obj = self._initialize_h5coro()
+        self.promise = self._read_datasets()
+
+    # TODO: Add http driver option
+    def _initialize_h5coro(self) -> h5coro.H5Coro:
+        """Initialize the h5coro object with the provided uri and credentials."""
+        if self.uri.startswith('s3://') and not self.credentials:
+            raise ValueError("AWS credentials are required for S3 URLs.")
+        
+        args = dict(
+            errorChecking=True,
+            verbose=False
+        )
+        if self.uri.startswith('s3://') and self.credentials:
+            args['resource'] = self.uri.replace('s3://', '')
+            args['driverClass'] = s3driver.S3Driver
+            args['credentials'] = dict(self.credentials)
+        else:
+            args['resource'] = self.uri
+            args['driverClass'] = filedriver.FileDriver
+        return h5coro.H5Coro(**args)
+
+    def _read_datasets(self) -> dict:
+        """Read datasets from the HDF5 file using h5coro."""
+        variable_paths = self.datasets.variables
+        dataset_objs = [dict(v) for k, v in variable_paths.items()] + [dict(self.datasets.longitude), dict(self.datasets.latitude)]
+        return self.h5obj.readDatasets(dataset_objs, block=False)
+
+    def _create_mask(self) -> np.ndarray:
+        """Create a mask for the datasets if masking is requested."""
+        if not self.mask_using:
+            return None
+        
+        mask_dataset_path = self.datasets.variables[self.mask_using].dataset
+        mask_dataset_points = self.promise[mask_dataset_path]
+        attributes = self.h5obj.list(mask_dataset_path)[1]
+        fill_value = attributes['_FillValue']
+        mask = (mask_dataset_points != fill_value)
+        
+        return mask if mask.any() else None
+
+    def _process_datasets(self, mask: np.ndarray):
+        """Process each dataset, apply mask if necessary, and create pyarrow arrays and schema fields."""
+        schema_fields = []
+        pyarrow_arrays = {}
+
+        for k, v in self.datasets.variables.items():
+            if k == self.mask_using:
+                continue
+            
+            variable_points = self.promise[v.dataset]
+            if mask is not None:
+                variable_points = variable_points[mask]
+            
+            schema_fields.append(pa.field(k, pa.from_numpy_dtype(variable_points.dtype)))
+            pyarrow_arrays[k] = pa.array(variable_points)
+
+        if self.mask_using:
+            mask_dataset_path = self.datasets.variables[self.mask_using].dataset
+            mask_dataset_points = self.promise[mask_dataset_path][mask]
+            schema_fields.append(pa.field(self.mask_using, pa.from_numpy_dtype(mask_dataset_points.dtype)))
+            pyarrow_arrays[self.mask_using] = pa.array(mask_dataset_points)
+
+        return pyarrow_arrays, schema_fields
```

Comment:
If you're going to create arrays and fields, it might be simpler to return a single logical `Table` or `RecordBatch` object instead. (A `RecordBatch` ensures that every column is in contiguous memory, whereas a `Table` has `ChunkedArrays` as columns)

You could do `pa.record_batch(pyarrow_arrays, schema=pa.schema(schema_fields))`